File: apps/backend/app/models/workflow.py
"""
Workflow data models for One Data Studio Lite

Defines database models for DAGs, nodes, edges, and executions.
"""

from datetime import datetime
from typing import Optional, List, Dict, Any
from sqlalchemy import (
    Column,
    Integer,
    String,
    Text,
    DateTime,
    JSON,
    ForeignKey,
    Enum as SQLEnum,
    Boolean,
    Float,
)
from sqlalchemy.orm import relationship

from app.core.database import Base


class DAG(Base):
    """
    DAG (Directed Acyclic Graph) model

    Represents a workflow with nodes and edges.
    """

    __tablename__ = "dags"

    id = Column(Integer, primary_key=True, index=True)
    dag_id = Column(String(250), unique=True, nullable=False, index=True)
    name = Column(String(500), nullable=False)
    description = Column(Text, nullable=True)

    # Scheduling
    schedule_interval = Column(String(100), nullable=True)  # Cron expression or None
    start_date = Column(DateTime, nullable=True)
    catchup = Column(Boolean, nullable=False, default=False)
    max_active_runs = Column(Integer, nullable=False, default=1)
    concurrency = Column(Integer, nullable=False, default=16)

    # State
    is_active = Column(Boolean, nullable=False, default=True)
    is_paused = Column(Boolean, nullable=False, default=False)

    # Metadata
    tags = Column(JSON, nullable=True, default=list)  # List of tag strings
    owner_id = Column(Integer, ForeignKey("users.id"), nullable=True, index=True)
    project_id = Column(Integer, ForeignKey("projects.id"), nullable=True, index=True)

    # Timestamps
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)
    last_parsed = Column(DateTime, nullable=True)

    # File info (for DAGs created from files)
    fileloc = Column(String(500), nullable=True)
    filepath = Column(String(500), nullable=True)

    # Relationships
    owner = relationship("User", back_populates="dags")
    project = relationship("Project", back_populates="dags")
    nodes = relationship("DAGNode", back_populates="dag", cascade="all, delete-orphan")
    edges = relationship("DAGEdge", back_populates="dag", cascade="all, delete-orphan")
    runs = relationship("DAGRun", back_populates="dag", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<DAG(id={self.id}, dag_id='{self.dag_id}', schedule='{self.schedule_interval}')>"


class DAGNode(Base):
    """
    DAG Node model

    Represents a single task/node in a DAG.
    """

    __tablename__ = "dag_nodes"

    id = Column(Integer, primary_key=True, index=True)
    dag_id = Column(Integer, ForeignKey("dags.id"), nullable=False, index=True)
    node_id = Column(String(250), nullable=False, index=True)
    name = Column(String(500), nullable=False)
    node_type = Column(String(50), nullable=False)  # sql, python, etl, training, etc.

    # Task configuration
    config = Column(JSON, nullable=True, default=dict)

    # Execution settings
    retry_count = Column(Integer, nullable=False, default=0)
    retry_delay_seconds = Column(Integer, nullable=False, default=300)
    timeout_seconds = Column(Integer, nullable=True)
    execution_timeout_seconds = Column(Integer, nullable=True)

    # Dependencies (calculated from edges)
    depends_on_ids = Column(JSON, nullable=True, default=list)  # List of node IDs

    # Position for visualization
    x_position = Column(Float, nullable=True)
    y_position = Column(Float, nullable=True)

    # Timestamps
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    dag = relationship("DAG", back_populates="nodes")
    outgoing_edges = relationship("DAGEdge", foreign_keys="[DAGEdge.source_node_id]", back_populates="source_node")
    incoming_edges = relationship("DAGEdge", foreign_keys="[DAGEdge.target_node_id]", back_populates="target_node")

    def __repr__(self):
        return f"<DAGNode(id={self.id}, node_id='{self.node_id}', type='{self.node_type}')>"


class DAGEdge(Base):
    """
    DAG Edge model

    Represents a dependency relationship between two nodes.
    """

    __tablename__ = "dag_edges"

    id = Column(Integer, primary_key=True, index=True)
    dag_id = Column(Integer, ForeignKey("dags.id"), nullable=False, index=True)
    source_node_id = Column(Integer, ForeignKey("dag_nodes.id"), nullable=False, index=True)
    target_node_id = Column(Integer, ForeignKey("dag_nodes.id"), nullable=False, index=True)

    # Edge condition (optional - for conditional dependencies)
    condition = Column(String(500), nullable=True)

    # Timestamps
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)

    # Relationships
    dag = relationship("DAG", back_populates="edges")
    source_node = relationship("DAGNode", foreign_keys=[source_node_id], back_populates="outgoing_edges")
    target_node = relationship("DAGNode", foreign_keys=[target_node_id], back_populates="incoming_edges")

    def __repr__(self):
        return f"<DAGEdge(id={self.id}, source={self.source_node_id}, target={self.target_node_id})>"


class DAGRun(Base):
    """
    DAG Run model

    Represents a single execution of a DAG.
    """

    __tablename__ = "dag_runs"

    id = Column(Integer, primary_key=True, index=True)
    dag_id = Column(Integer, ForeignKey("dags.id"), nullable=False, index=True)
    run_id = Column(String(250), unique=True, nullable=False, index=True)

    # Execution info
    execution_date = Column(DateTime, nullable=False, index=True)
    start_date = Column(DateTime, nullable=True)
    end_date = Column(DateTime, nullable=True)

    # State
    state = Column(
        SQLEnum("queued", "running", "success", "failed", "cancelled", "paused", name="dag_run_state"),
        nullable=False,
        default="queued",
        index=True,
    )

    # Run configuration
    conf = Column(JSON, nullable=True, default=dict)
    run_type = Column(String(50), nullable=False, default="manual")  # manual, scheduled, backfill

    # Execution stats
    duration = Column(Float, nullable=True)  # Duration in seconds
    max_tries = Column(Integer, nullable=False, default=0)

    # User info
    triggered_by = Column(Integer, ForeignKey("users.id"), nullable=True)

    # Timestamps
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    dag = relationship("DAG", back_populates="runs")
    trigger = relationship("User", foreign_keys=[triggered_by])
    task_instances = relationship("TaskInstance", back_populates="dag_run", cascade="all, delete-orphan")

    def __repr__(self):
        return f"<DAGRun(id={self.id}, run_id='{self.run_id}', state='{self.state}')>"


class TaskInstance(Base):
    """
    Task Instance model

    Represents a single task execution within a DAG run.
    """

    __tablename__ = "task_instances"

    id = Column(Integer, primary_key=True, index=True)
    dag_run_id = Column(Integer, ForeignKey("dag_runs.id"), nullable=False, index=True)
    node_id = Column(Integer, ForeignKey("dag_nodes.id"), nullable=False, index=True)
    task_id = Column(String(250), nullable=False, index=True)  # Unique identifier for this task instance

    # Execution info
    start_date = Column(DateTime, nullable=True)
    end_date = Column(DateTime, nullable=True)
    duration = Column(Float, nullable=True)  # Duration in seconds

    # State
    state = Column(
        SQLEnum(
            "pending",
            "queued",
            "running",
            "success",
            "failed",
            "skipped",
            "upstream_failed",
            "retried",
            name="task_instance_state",
        ),
        nullable=False,
        default="pending",
        index=True,
    )

    # Retry info
    try_number = Column(Integer, nullable=False, default=1)
    max_tries = Column(Integer, nullable=False, default=0)
    previous_try_task_id = Column(Integer, ForeignKey("task_instances.id"), nullable=True)

    # Result
    result = Column(JSON, nullable=True)  # Task output
    error = Column(Text, nullable=True)  # Error message if failed

    # Host/Executor info
    hostname = Column(String(500), nullable=True)
    executor = Column(String(50), nullable=True)  # e.g., "LocalExecutor", "KubernetesExecutor"

    # Logs
    log_url = Column(String(1000), nullable=True)
    log_filepath = Column(String(1000), nullable=True)

    # Timestamps
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)
    updated_at = Column(DateTime, nullable=False, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    dag_run = relationship("DAGRun", back_populates="task_instances")
    node = relationship("DAGNode")
    previous_try = relationship("TaskInstance", foreign_keys=[previous_try_task_id])

    def __repr__(self):
        return f"<TaskInstance(id={self.id}, task_id='{self.task_id}', state='{self.state}')>"

File: apps/frontend/src/types/workflow.ts
/**
 * Workflow Type Definitions
 */

// Node position
export interface Position {
  x: number;
  y: number;
}

// DAG Node configuration
export interface NodeConfig {
  description?: string;
  retry_count?: number;
  retry_delay_seconds?: number;
  timeout_seconds?: number;
  parameters?: Record<string, any>;
  depends_on?: string[];
  // Task-type specific configs
  sql?: string;
  conn_id?: string;
  pipeline_id?: string;
  experiment_id?: string;
  command?: string;
  code?: string;
  image?: string;
  resources?: NodeResources;
}

// Node resource requirements
export interface NodeResources {
  cpu?: string;
  memory?: string;
  gpu?: number;
  gpu_type?: string;
}

// DAG Node
export interface DAGNode {
  id: string;
  task_type: TaskTypeName;
  name: string;
  description?: string;
  config?: NodeConfig;
  position: Position;
  status?: 'pending' | 'running' | 'success' | 'failed' | 'skipped' | 'upstream_failed';
}

// Task type names
export type TaskTypeName =
  | 'sql'
  | 'python'
  | 'shell'
  | 'etl'
  | 'training'
  | 'inference'
  | 'evaluation'
  | 'model_register'
  | 'wait'
  | 'sensor'
  | 'email'
  | 'webhook'
  | 'slack'
  | 'export'
  | 'import'
  | 'notebook';

// DAG Edge
export interface DAGEdge {
  id: string;
  source: string;
  target: string;
  condition?: string;
}

// DAG
export interface DAG {
  id: number;
  dag_id: string;
  name: string;
  description?: string;
  schedule_interval?: string;
  is_active: boolean;
  is_paused: boolean;
  tags: string[];
  owner_id?: number;
  created_at: string;
  updated_at: string;
}

// DAG Run
export interface DAGRun {
  id: number;
  dag_id: number;
  run_id: string;
  execution_date: string;
  state: 'queued' | 'running' | 'success' | 'failed' | 'cancelled' | 'paused';
  start_date?: string;
  end_date?: string;
  run_type: string;
  duration?: number;
}

// Task Instance
export interface TaskInstance {
  id: number;
  task_id: string;
  node_id: number;
  state: 'pending' | 'queued' | 'running' | 'success' | 'failed' | 'skipped' | 'upstream_failed' | 'retried';
  start_date?: string;
  end_date?: string;
  duration?: number;
  try_number: number;
  max_tries: number;
  error?: string;
  log_url?: string;
}

// Task Type
export interface TaskType {
  type: TaskTypeName;
  name: string;
  category: TaskCategory;
  description?: string;
  icon?: string;
  color?: string;
}

// Task Categories
export type TaskCategory =
  | 'Data'
  | 'Code'
  | 'Machine Learning'
  | 'Control Flow'
  | 'Notification'
  | 'Data Transfer'
  | 'Notebook'
  | 'Integration';

// Task Type Category mapping
export const TASK_TYPE_CATEGORIES: Record<TaskCategory, TaskTypeName[]> = {
  Data: ['sql', 'etl'],
  Code: ['python', 'shell'],
  'Machine Learning': ['training', 'inference', 'evaluation', 'model_register'],
  'Control Flow': ['wait', 'sensor'],
  Notification: ['email', 'webhook', 'slack'],
  'Data Transfer': ['export', 'import'],
  Notebook: ['notebook'],
  Integration: [],
};

// Workflow Template
export interface WorkflowTemplate {
  id: string;
  name: string;
  description: string;
  category: string;
  icon?: string;
  tags?: string[];
  tasks: TemplateTask[];
  variables?: TemplateVariable[];
  thumbnail?: string;
}

// Template Task
export interface TemplateTask {
  task_id: string;
  task_type: TaskTypeName;
  name: string;
  description?: string;
  depends_on?: string[];
  parameters?: Record<string, any>;
  position?: Position;
}

// Template Variable
export interface TemplateVariable {
  name: string;
  type: 'string' | 'number' | 'boolean' | 'select' | 'multiline';
  label: string;
  default?: any;
  required?: boolean;
  options?: string[];
  description?: string;
}

// Workflow Export Format
export interface WorkflowExport {
  version: string;
  exported_at: string;
  dag: {
    dag_id: string;
    name: string;
    description?: string;
    schedule_interval?: string;
    tags: string[];
    tasks: TemplateTask[];
    edges?: DAGEdge[];
  };
}

// Canvas State
export interface CanvasState {
  scale: number;
  position: Position;
  minScale: number;
  maxScale: number;
}

// Selection State
export interface SelectionState {
  nodes: string[];
  edges: string[];
}

// Validation Error
export interface ValidationError {
  type: 'node' | 'edge' | 'dag';
  id?: string;
  message: string;
  severity: 'error' | 'warning';
}

// ETL Pipeline Reference
export interface ETLPipelineRef {
  id: string;
  name: string;
  description?: string;
  source_type: string;
  target_type: string;
  step_count: number;
}

// Experiment Reference
export interface ExperimentRef {
  id: string;
  name: string;
  description?: string;
  project_id?: string;
}

// Model Reference
export interface ModelRef {
  id: string;
  name: string;
  version: string;
  stage?: 'development' | 'staging' | 'production' | 'archived';
}

// Scheduling Options
export interface ScheduleOptions {
  cron_expression?: string;
  start_date?: string;
  end_date?: string;
  catchup?: boolean;
  max_active_runs?: number;
  concurrency?: number;
}

// Notification Config
export interface NotificationConfig {
  on_success?: boolean;
  on_failure?: boolean;
  on_retry?: boolean;
  channels: {
    email?: {
      to: string[];
      subject?: string;
    };
    slack?: {
      webhook: string;
      channel?: string;
    };
    webhook?: {
      url: string;
      method?: 'POST' | 'PUT' | 'PATCH';
      headers?: Record<string, string>;
    };
  };
}

// Workflow Execution History
export interface ExecutionHistory {
  run_id: string;
  execution_date: string;
  state: string;
  duration?: number;
  triggered_by?: string;
  tasks: {
    total: number;
    success: number;
    failed: number;
    running: number;
    pending: number;
  };
}

// Node execution result
export interface NodeExecutionResult {
  node_id: string;
  state: string;
  start_time: string;
  end_time?: string;
  duration?: number;
  output?: any;
  error?: string;
  logs?: string[];
}

// Drag and Drop Types
export interface DragItem {
  type: string;
  taskType: TaskType;
  position?: Position;
}

// View Mode
export type ViewMode = 'edit' | 'view' | 'monitor';
